### PR TITLE
feat:support dubbo.load()

### PR DESCRIPTION
**Pull request by FinalT**

### CODE REVIEW

File: dubbo.go [hunk 1]
```diff
@@ -137,3 +145,63 @@ func (ins *Instance) NewServer(opts ...server.ServerOption) (*server.Server, err
 	}
 	return srv, nil
 }
+
+func (ins *Instance) Start() (err error) {
```
Comment (DMwangnima) [inline dubbo.go:149]: There is no need to expose this method?

File: dubbo.go [hunk 2]
```diff
@@ -137,3 +145,63 @@ func (ins *Instance) NewServer(opts ...server.ServerOption) (*server.Server, err
...
+		if err = ins.LoadProvider(); err != nil {
+			return
+		}
+	})
+	return err
+}
+
+// LoadProvider loads the service provider.
+func (ins *Instance) LoadProvider() error {
+	srv, err := ins.NewServer()
```
Comment (DMwangnima) [inline dubbo.go:164]: Just consider what options should be passed to ```NewServer``` and ```srv.Register``` referring to ```ProviderConfig``` and ```ReferenceConfig```.

File: dubbo.go [hunk 3]
```diff
@@ -137,3 +145,63 @@ func (ins *Instance) NewServer(opts ...server.ServerOption) (*server.Server, err
...
+		if err = srv.Register(definition.Handler, definition.Info, definition.Opts...); err != nil {
+			return err
+		}
+	}
+	return srv.Start()
+}
+
+// LoadConsumer loads the service consumer.
+func (ins *Instance) LoadConsumer() error {
+	cli, err := ins.NewClient()
```
Comment (DMwangnima) [inline dubbo.go:179]: ditto. Refer to ```ConsumerConfig``` and "ReferenceConfig".

File: dubbo.go [hunk 4]
```diff
@@ -137,3 +145,63 @@ func (ins *Instance) NewServer(opts ...server.ServerOption) (*server.Server, err
...
+	if err != nil {
+		return err
+	}
+	// refer services
+	for _, definition := range consumerServices {
+		if err = cli.Init(definition.Info); err != nil {
+			return err
+		}
+		definition.Info.ClientInjectFunc(definition.Svc, cli)
```
Comment (DMwangnima) [inline dubbo.go:188]: Excellent!

File: server/server.go [hunk 5]
```diff
@@ -181,6 +182,15 @@ func (s *Server) Serve() error {
 	select {}
 }
 
+func (s *Server) Start() error {
+	if err := s.exportServices(); err != nil {
+		return err
+	}
+	metadata.ExportMetadataService()
+	registry_exposed.RegisterServiceInstance(s.cfg.Application.Name, s.cfg.Application.Tag, s.cfg.Application.MetadataType)
```
Comment (DMwangnima) [inline server/server.go:190]: This part is used for what?

File: dubbo.go [hunk 6]
```diff
@@ -137,3 +145,63 @@ func (ins *Instance) NewServer(opts ...server.ServerOption) (*server.Server, err
...
+		}
+		if err = ins.LoadProvider(); err != nil {
+			return
+		}
+	})
+	return err
+}
+
+// LoadProvider loads the service provider.
+func (ins *Instance) LoadProvider() error {
```
Comment (DMwangnima) [inline dubbo.go:163]: ditto.

File: dubbo.go [hunk 7]
```diff
@@ -137,3 +145,63 @@ func (ins *Instance) NewServer(opts ...server.ServerOption) (*server.Server, err
...
+	for _, definition := range providerServices {
+		if err = srv.Register(definition.Handler, definition.Info, definition.Opts...); err != nil {
+			return err
+		}
+	}
+	return srv.Start()
+}
+
+// LoadConsumer loads the service consumer.
+func (ins *Instance) LoadConsumer() error {
```
Comment (DMwangnima) [inline dubbo.go:178]: ditto.

File: server/server.go (hunk 5 above)
Comment (chickenlj) [reply server/server.go:190]: I have the same question, we already have `server.Serve()`. And also `dubbo.Start`, `dubbo.LoadProvider`,`dubbo.LoadConsumer` are not necesary.

File: dubbo.go (hunk 1 above)
Comment (FinalT) [reply dubbo.go:149]: Yeah, I don't think it needs to be exposed

File: server/server.go (hunk 5 above)
Comment (FinalT) [reply server/server.go:190]: Non-blocking start, same as the original

Comment (FinalT) [reply server/server.go:190]: I need a func to start without blocking

Comment (chickenlj) [reply server/server.go:190]: How about you we make the new method private and delegate the main logic of `Serve` to this method by calling `ins.start()`?

Comment (DMwangnima) [reply server/server.go:190]: ```config.Load``` is non-blocking. We could put ```server.Serve()``` in a goroutine to make ```dubbo.Load()``` non-blocking. Then there would be enough space for us to consider whether introducing a non-blocking method ```server.Start()``` is necessary.

**Pull request closed without merging**